snow/validators/manager.go

```diff
@@ -22,25 +23,83 @@ var (
 	ErrMissingValidators = errors.New("missing validators")
 )
 
+type SetCallbackListener interface {
+	OnValidatorAdded(validatorID ids.NodeID, pk *bls.PublicKey, txID ids.ID, weight uint64)
+	OnValidatorRemoved(validatorID ids.NodeID, weight uint64)
+	OnValidatorWeightChanged(validatorID ids.NodeID, oldWeight, newWeight uint64)
+}
+
 // Manager holds the validator set of each subnet
 type Manager interface {
 	fmt.Stringer
 
-	// Add a subnet's validator set to the manager.
-	//
-	// If the subnet had previously registered a validator set, false will be
-	// returned and the manager will not be modified.
-	Add(subnetID ids.ID, set Set) bool
+	// Add a new staker to the subnet.
+	// Returns an error if:
+	// - [weight] is 0
+	// - [nodeID] is already in the validator set
+	// If an error is returned, the set will be unmodified.
+	AddStaker(subnetID ids.ID, nodeID ids.NodeID, pk *bls.PublicKey, txID ids.ID, weight uint64) error
+
+	// AddWeight to an existing staker to the subnet.
+	// Returns an error if:
+	// - [weight] is 0
+	// - [nodeID] is not already in the validator set
+	// If an error is returned, the set will be unmodified.
+	// AddWeight can result in a total weight that overflows uint64.
+	// In this case no error will be returned for this call.
+	// However, the next TotalWeight call will return an error.
+	AddWeight(subnetID ids.ID, nodeID ids.NodeID, weight uint64) error
+
+	// GetWeight retrieves the validator weight from the subnet.
+	GetWeight(subnetID ids.ID, validatorID ids.NodeID) uint64
+
+	// GetValidator returns the validator tied to the specified ID in subnet.
+	// If the validator doesn't exist, returns false.
+	GetValidator(subnetID ids.ID, validatorID ids.NodeID) (*Validator, bool)
```

Comment:
TL;DR GetValidator does not seem used in avalanchego. If I comment it out (and overridden_manager too), the binary gets build. Can we drop this method?
@ceyonur  more generally, this has become a huge interface whose methods seems kind of duplicated. I understand there are efficiency reason to have them (GetValidator does copy the Validator struct to ensure isolation, while e.g. GetWeigh does not need to copy). Still we should look for ways to reduce the interface, or at least fragment it?

Reply:
I feel like this method is a very core (bare minimum) function for this interface. Even if we don't use it, we should keep this in the interface (to serve as an external library).